Python/import.c

```diff
@@ -303,10 +303,116 @@ _PyImport_Fini(void)
+    if (modules == NULL)
+        Py_FatalError("lost sys.modules");
+    return modules;
+}
+
+/* In some corner cases it is important to be sure that the import
+   machinery has been initialized (or not cleaned up yet).  For
+   example, see issue #4236 and PyModule_Create2(). */
+
+int
+_PyImport_IsInitialized(PyInterpreterState *interp)
+{
+    if (interp->sysdict == NULL)
+        return 0;
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(interp->sysdict, &PyId_modules);
+    if (modules == NULL)
+        return 0;
+    return 1;
+}
+
+PyObject *
+_PyImport_GetModule(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
+        return PyDict_GetItem(modules, name);
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior of PyDict_GetItem().
+    if (PyErr_Occurred())
+        PyErr_Clear();
+    Py_XDECREF(mod);
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleWithError(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
+        return PyDict_GetItemWithError(modules, name);
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior
+    // of PyDict_GetItemWithError().
+    if (PyErr_Occurred() && !PyMapping_HasKey(modules, name))
+        PyErr_Clear();
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleString(const char *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
+        return PyDict_GetItemString(modules, name);
+
+    PyObject *mod = PyMapping_GetItemString(modules, name);
+    // For backward-comaptibility we copy the behavior
+    // of PyDict_GetItemString().
+    if (PyErr_Occurred())
+        PyErr_Clear();
+    Py_XDECREF(mod);
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleId(struct _Py_Identifier *nameid)
+{
+    PyObject *name = _PyUnicode_FromId(nameid); /* borrowed */
+    if (name == NULL)
+        return NULL;
+    return _PyImport_GetModule(name);
+}
+
+int
+_PyImport_SetModule(PyObject *name, PyObject *m) {
+    PyObject *modules = PyImport_GetModuleDict();
+    return PyObject_SetItem(modules, name, m);
+}
+
+int
+_PyImport_SetModuleString(const char *name, PyObject *m) {
+    PyObject *modules = PyImport_GetModuleDict();
+    return PyMapping_SetItemString(modules, name, m);
+}
+
+PyObject *
+PyImport_GetModule(PyObject *name)
+{
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PySys_GetObjectId(&PyId_modules);
+    if (modules == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "unable to get sys.modules");
+        return NULL;
+    }
+    if (PyDict_Check(modules))
+        return PyDict_GetItemWithError(modules, name);
+
+    PyObject *m = PyObject_GetItem(modules, name);
+    if (PyErr_Occurred() && !PyMapping_HasKey(modules, name))
+        PyErr_Clear();
+    return m;
```

Comment:
Yes, part of the point here is to support arbitrary mappings (anything you might assign to sys.modules).